pkg/kv/kvserver/rafttermcache/term_cache.go

```diff
@@ -0,0 +1,384 @@
+	//
+	// 	From index 14 to 18, Note (14, t2) , (18, t4) in the raftLog,
+	// 	we jumped from term2 to term4.
+	// 	Due to the properties of the termCache, we know there is no entries from
+	// 	index 14 to 18 that has term 3.
+	// 	The raftLog must be: [14/t2, 15/t2, 16/t2, 17/t2, 18/t4]
+	// 	Because we update the termCache by examining each new entry.
+	// 	We modify the cache by appending to the end if there is a term flip entry.
+	// 	We never modify the cache by inserting into the middle.
+	//  If the cache is full, we evict the earliest entry.
+	// 	So there are no entries with term = 3 in the raftLog.
+	//
+	// 	Similarly, from index 18 to 22:
+	// 	The raftLog must be: [18/t4, 19/t4, 20/t4, 21/t4, 22/t4, 23/t6]
+	// 	So there are no entries with term = 5 in the raftLog,
+	//
+	// 	Term 6 is the last entry in cache.
+	// 	Since we also know the lastIndex covered by termCache:
+	// 	Therefore from index 22 to 25:
+	// 	The raftLog must be: [22/t6, 23/t6, 24/t6, 25/t6]
+	// TODO(hakuuww): consider using a more efficient data structure
+	//  that doesn't require copying slices (ie. circular buffer)
+	//  If we append entries in big batches, and term flips somehow happen
+	//  extremely frequently, we will be bottlenecked by the term cache
+	//  in the extreme case.
+	cache []entryID
+	// maxSize is the maximum size of the term cache.
+	maxSize uint64
+	// lastIndex is the last index known to the TermCache that is in the raftLog.
+	// the entry at lastIndex has the same term as TermCache.cache's
+	// last entry's term.
+	//
+	// lastIndex is set to 0 when TermCache is created.
+	//
+	// lastIndex is first updated with the storage engine's
+	// term: r.shMu.lastTermNotDurable and index: r.shMu.lastIndexNotDurable
+	// as the first entry. This entry is the last entry in the raftLog's
+	// persisted part.
+	//
+	// in the special case when raftLog is completely empty,
+	// (can happen when deploying the database for the first time with no data)
+	// the TruncatedState == LastEntryID
+	// (the TruncatedState represents the highest applied-to-state machine entry)
+	//
+	//
+	// lastIndex is updated when entries are appended to the TermCache.
+	//
+	// Invariants (except for on the creation of the TermCache):
+	//
+	// For callers, lastIndex is always populated (lastIndex >= 0),
+	// and there is always at least one entry being covered in the termCache.
+	//
+	// lastIndex mirrors the last index in the storage part of raftLog.
+	lastIndex kvpb.RaftIndex
+	// mu is used to protect the term cache from concurrent accesses.
+	// Which can happen for example when the following is called concurrently:
+	// - logstore.LoadTerm() in replicaLogStorage.termLocked(),
+	// - ScanAppend() in logstore/storeEntriesAndCommitBatch().
+	// TODO(hakuuww): using a RWMutex is one way, and can guarantee safeness.
+	//  But maybe we can just lock replica.mu externally when calling ScanAppend()
+	//  and avoid using locks when calling LoadTerm().
+	mu syncutil.RWMutex
```

Comment:
After moving to `raft` package, the mutex/synchronization is no longer needed. All access inside the `RawNode` is guarded by `Replica.mu`.